runtime/commonMain/src/kotlinx/serialization/Annotations.kt

```diff
@@ -10,60 +10,156 @@ import kotlinx.serialization.json.*
 import kotlin.reflect.*
 
 /**
- * Sets specific serializer for class, property or type argument.
- * When argument is omitted, plugin will generate default implementation inside the class.
+ * The main entry point to the serialization process.
+ * Applying [Serializable] to the Kotlin class instructs the serialization plugin to automatically generate [KSerializer]
+ * for the current class, that can be used to serialize and deserialize the class.
+ * The generated serializer can be accessed with `.serializer()` extension method that is generated by the plugin as well.
+ * ```
+ * @Serializable
+ * class MyData(val myData: AnotherData, val intProperty: Int, ...)
+ *
+ * // Produces JSON string using the generated serializer
+ * val jsonString = Json.encodeToJson(MyData.serializer(), instance)
+ * ```
+ *
+ * Additionally, the user-defined serializer can be specified using [with] parameter:
+ * ```
+ * @Serializable(with = MyAnotherDataCustomSerializer::class)
+ * class MyAnotherData(...)
+ *
+ * MyAnotherData.serializer() // <- returns MyAnotherDataCustomSerializer
+ * ```
+ *
+ * For annotated properties, specifying [with] parameter is mandatory and can be used to override
+ * serializer on the use-site without affecting the rest of the usages:
+ * ```
+ * @Serializable // By default is serialized as 3 byte components
+ * class RgbPixel(val red: Byte, val green: Byte, val blue: Byte)
+ *
+ * @Serializable
+ * class RgbExample(
+ *     @Serializable(with = RgbAsHexString::class) p1: RgpPixel, // Serialize as HEX string, e.g. #FFFF00
+ *     @Serializable(with = RgbAsSingleInt::class) p2: RgpPixel, // Serialize as single integer, e.g. 16711680
+ *     p3: RgpPixel // Serialize as 3 byte components, e.g. { "red": 255, "green": 255, "blue": 0 }
+ * )
+ * ```
+ * In this example, each pixel will be serialized using different data representation.
+ *
+ * For classes with generic type parameters, `serializer()` method requires one additional argument per each generic type parameter:
+ * ```
+ * @Serializable
+ * class Box<T>(value: T)
+ *
+ * Box.serializer() // Doesn't compile
+ * Box.serializer(Int.serializer()) // Returns serializer for Box<Int>
+ * Box.serializer(Box.serializer(Int.serializer())) // Returns serializer for Box<Box<Int>>
+ * ```
+ *
+ * ### Implementation details
+ *
+ * In order to generate `serializer` method that is not a method on the particular instance, the class should have a companion object, either named or unnamed.
+ * Companion object is generated by the plugin if it is not declared, effectively exposing both companion and `serializer()` method to class ABI.
+ * If companion object already exists, only `serializer` method will be generated.
+ * If `serializer` method already exists, it will be matched against the desired signature `serializer(vararg type params): KSerializer<T>`.
```

Comment:
`serializer(vararg type params)` -- we do not have such signature. generated serializer() function has exactly N KSerializer parameters, where N is number of type parameters.